### PR TITLE
user invites: Add home page invite colleagues feature

### LINKED ISSUE #31703: user invites on-prem: Add site or user configuration option to turn off user invitations on on-prem
Part of https://github.com/sourcegraph/sourcegraph/issues/31214

See https://sourcegraph.slack.com/archives/C01LZKLRF0C/p1645585709938119?thread_ts=1645565266.889989&cid=C01LZKLRF0C for more context

### PR DESCRIPTION
**Pull request by philipp-spiess**
Fixes https://github.com/sourcegraph/sourcegraph/issues/31558
Fixes #31703

This is a first iteration on the UI for the changes explained in https://github.com/sourcegraph/sourcegraph/issues/31558. 

We want to allow users to invite potential collaborators when they have finished the post-sign-up process and on on-prem.

We're [still working on further improving the messaging for users and admins](https://sourcegraph.slack.com/archives/C01LZKLRF0C/p1645665484415589), so this is not intended to be shipped this way (hence I put it behind the experimental feature flag). I wanted to get this up for review already though since it slowly grows and it's better to merge into `main` to avert further diversion.

TL;DR of my changes:

- We add a new feature flag that, when enabled _and_ the instance has emails configured, will change the saved searches/community contexts section into a tab panel and display the invitations panel there as well.
- The invitations panel was made so that it works in all screen sizes. Individual users shown there are buttons and it will show the invite text when hovering or focusing via keyboard navigation.
- The tab panel will remember the selection so a user can opt-out of this view by switching to the other tab just ones.
- The test plan contains screenshots to make sure nothing regresses when the feature flag is disabled.
- Most of the meat here was about getting the layout to work correctly, especially in responsive mode. 

## Test plan

I plan on adding UI tests later as we have ruled out some of the edge cases here and are closer to opening up the feature flag so for now you have to suffice with a collection of screenshots and videos.

### [Test Matrix](https://github.com/philipp-spiess/philipp-spiess/blob/main/Engineering/Test%20Matrixes.md)

|                         | Cloud | On-Prem |
|-------------------------|-------|---------|
| Disabled                |    cloud-disabled   |     onprem-disabled    |
| **Mail not configured** |   cloud-no-email-enabled---NOTE    |     onprem-no-email-enabled---NOTE    |
| No contributors         |     cloud-no-contributors  |      onprem-no-contributors   |
| Recommendation          |   cloud-recommendation    |        onprem-recommendation |

__Note: When Emails are not configured, we might want to change the text cc @rrhyne.__

### Functionality

https://user-images.githubusercontent.com/458591/156162865-9b78c199-10fd-4450-bf8b-f6678835ed0c.mov

### Responsive Mode

https://user-images.githubusercontent.com/458591/156162709-45327481-87a3-4d8c-85b1-95b3fc3db26d.mov

### CODE REVIEW

File: client/web/src/SourcegraphWebApp.tsx [hunk 1]
```diff
@@ -460,6 +460,7 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
                                                                 }
                                                                 fetchSavedSearches={fetchSavedSearches}
                                                                 fetchRecentSearches={fetchRecentSearches}
+                                                                fetchCollaborators={fetchCollaborators}
                                                                 fetchRecentFileViews={fetchRecentFileViews}
```
Comment (limitedmage) [inline client/web/src/SourcegraphWebApp.tsx:464]: Not necessarily something to do in this PR, but we should consider in the near future refactoring this to a) use Apollo b) cache results c) get all data required to load panels in a single GraphQL call instead of 4-5 separate calls. All of this should help performance.

File: client/web/src/search/backend.tsx [hunk 2]
```diff
@@ -268,3 +272,22 @@ export function fetchRecentSearches(userId: Scalars['ID'], first: number): Obser
 export function fetchRecentFileViews(userId: Scalars['ID'], first: number): Observable<EventLogResult | null> {
     return fetchEvents(userId, first, 'ViewBlob')
 }
+
+export function fetchCollaborators(userId: Scalars['ID']): Observable<InvitableCollaborator[]> {
+    if (!userId) {
+        return of([])
+    }
+
+    // @todo Change this API to use the new backend added as part of #31777
```
Comment (limitedmage) [inline client/web/src/search/backend.tsx:281]: Is there an issue for this or will this be added in this PR? IMO ToDo comments that are checked in should always have issues filed for them.

File: client/web/src/search/panels/SavedSearchesPanel.tsx [hunk 3]
```diff
@@ -147,12 +148,29 @@ export const SavedSearchesPanel: React.FunctionComponent<Props> = ({
...
+                    aria-label="Filter saved searches"
+                >
+                    ...
+                </MenuButton>
+
+                <MenuList>
+                    <MenuItem onSelect={() => setShowAllSearches(false)}>My searches</MenuItem>
+                    <MenuItem onSelect={() => setShowAllSearches(true)}>All searches</MenuItem>
+                </MenuList>
+            </Menu>
```
Comment (limitedmage) [inline client/web/src/search/panels/SavedSearchesPanel.tsx:166]: This new menu for small screens is awesome, thanks for adding it!

File: client/web/src/search/panels/SavedSearchesPanel.tsx [hunk 4]
```diff
@@ -147,12 +148,29 @@ export const SavedSearchesPanel: React.FunctionComponent<Props> = ({
...
+        </>
     )
     return (
         <PanelContainer
-            className={classNames(className, 'saved-searches-panel')}
+            hideTitle={hideTitle}
             title="Saved searches"
+            className={classNames(className, 'saved-searches-panel', { 'h-100': hideTitle })}
```
Comment (limitedmage) [inline client/web/src/search/panels/SavedSearchesPanel.tsx:173]: is the `'saved-searches-panel'` class still needed here?

File: client/web/src/search/panels/PanelContainer.tsx [hunk 5]
```diff
@@ -24,12 +25,19 @@ export const PanelContainer: React.FunctionComponent<Props> = ({
...
-            <h4 className={styles.headerText}>{title}</h4>
-            {actionButtons}
-        </div>
+        {hideTitle !== true ? (
+            <div className={classNames('d-flex border-bottom', styles.header)}>
+                <h4 className={styles.headerText}>{title}</h4>
+                {actionButtons}
+            </div>
+        ) : (
+            <div className={classNames('position-absolute', styles.header, styles.headerHiddenTitle)}>
```
Comment (limitedmage) [inline client/web/src/search/panels/PanelContainer.tsx:37]: `'position-absolute'` isn't needed here since it's already in `styles.headerHiddenTitle`

File: client/web/src/search/panels/PanelContainer.module.scss [hunk 6]
```diff
@@ -13,6 +13,12 @@
     }
 }
 
+.header-hidden-title {
+    position: absolute;
```
Comment (limitedmage) [inline client/web/src/search/panels/PanelContainer.module.scss:17]: I would recommend adding a comment that this `position:absolute` is specifically for relative position inside a tab panel. As part of this, renaming the prop `hideTitle` to `insideTabPanel` might make sense. Otherwise, a consumer of the panels might think that `hideTitle` only hides the title, when it also affects the display of any action buttons to be absolute positioned.

File: client/web/src/search/panels/HomePanels.tsx [hunk 7]
```diff
@@ -16,22 +18,69 @@ import { SavedSearchesPanel } from './SavedSearchesPanel'
...
-
-            {props.isSourcegraphDotCom ? (
-                <CommunitySearchContextsPanel {...props} className={classNames('col-lg-5', styles.panel)} />
-            ) : (
-                <SavedSearchesPanel {...props} className={classNames('col-lg-5', styles.panel)} />
-            )}
+const INVITES_TAB_KEY = 'HomePage.UserInvites.Tab'
+
+export const HomePanels: React.FunctionComponent<Props> = (props: Props) => {
+    const [persistedTabIndex, setPersistedTabIndex] = useLocalStorage(INVITES_TAB_KEY, 1)
```
Comment (limitedmage) [inline client/web/src/search/panels/HomePanels.tsx:27]: Suggestion: use [TemporarySettings](https://docs.sourcegraph.com/dev/background-information/web/temporary_settings) instead of localStorage here.  This way users will have their preferred tab synced across devices.

File: client/web/src/search/panels/HomePanels.tsx [hunk 8]
```diff
@@ -16,22 +18,69 @@ import { SavedSearchesPanel } from './SavedSearchesPanel'
 interface Props extends TelemetryProps, HomePanelsProps {
     authenticatedUser: AuthenticatedUser | null
     isSourcegraphDotCom: boolean
+    showCollaborators: boolean
```
Comment (limitedmage) [inline client/web/src/search/panels/HomePanels.tsx:21]: Suggestion: add storybook tests that set this to true and display the collaborators invites panel

File: client/web/src/search/panels/HomePanels.tsx [hunk 9]
```diff
@@ -16,22 +18,69 @@ import { SavedSearchesPanel } from './SavedSearchesPanel'
...
+                            </TabList>
+                            <TabPanels className="h-100">
+                                <TabPanel className="h-100">
+                                    <SavedSearchesPanel {...props} hideTitle={true} />
+                                </TabPanel>
+                                <TabPanel className="h-100">
+                                    <CollaboratorsPanel {...props} />
+                                </TabPanel>
+                            </TabPanels>
+                        </Tabs>
```
Comment (limitedmage) [inline client/web/src/search/panels/HomePanels.tsx:78]: minor, maybe overkill: can we extract this into a component since it's used twice here?

File: client/web/src/search/panels/CollaboratorsPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+interface Props extends TelemetryProps {
+    className?: string
+    authenticatedUser: AuthenticatedUser | null
+    fetchCollaborators: (userId: string) => Observable<InvitableCollaborator[]>
+}
+
+const emailEnabled = window.context?.emailEnabled ?? false
+
+export const CollaboratorsPanel: React.FunctionComponent<Props> = ({
```
Comment (limitedmage) [inline client/web/src/search/panels/CollaboratorsPanel.tsx:31]: Can some storybook tests be added for this component, in all its states (loading, empty, error)?

File: client/web/src/search/panels/CollaboratorsPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            className={classNames(
+                'd-flex',
+                'align-items-center',
+                'flex-column',
+                'justify-content-center',
+                'col-lg-12',
+                'h-100'
+            )}
+        >
+            <div className="text-center">No collaborators found in sampled repositories.</div>
```
Comment (limitedmage) [inline client/web/src/search/panels/CollaboratorsPanel.tsx:173]: What does "sampled" mean? How is the sample determined and can the user modify this sample?

File: client/web/src/search/panels/CollaboratorsPanel.tsx [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+            <div className="text-muted mt-3 text-center">
+                You can invite people to Sourcegraph with this direct link:
+            </div>
+            <CopyableText
+                className="mt-3"
+                text={inviteURL}
```
Comment (limitedmage) [inline client/web/src/search/panels/CollaboratorsPanel.tsx:179]: Will this link work if admin has direct registration disabled?

**Review by limitedmage — COMMENTED** [review]
This looks very good, great job! I left some overall suggestions and questions.

File: client/web/src/SourcegraphWebApp.tsx (hunk 1 above)
Comment (philipp-spiess) [reply client/web/src/SourcegraphWebApp.tsx:464]: Yeah agree I noticed this as well. Moreover these GraphQL queries currently only load when the respective panel components are mounted so we have to download quite a bit of JS before even starting to load data.

One thing to be mindful of here is that, if we hoist all data fetching up, we might quickly start to load more than we need. E.g. collaborators are only shown when the tab is active and ideally we would only load it as part of the initial page load when we know we show it. These logic is currently implemented on the client only though.

File: client/web/src/search/backend.tsx (hunk 2 above)
Comment (philipp-spiess) [reply client/web/src/search/backend.tsx:281]: Makes sense! I have a commit up that fixes this already but no issue, you're right.

**Review by limitedmage — APPROVED** [review]
